### ISSUE TITLE
Project with standard maven layout is not correctly imported

### ISSUE THREAD

**Issue by laeubi** [eclipse-m2e/m2e-core#1200]
An example could be found here:

https://github.com/eclipse-tycho/tycho/tree/master/demo/testing/tycho/samemodule

this is a standard maven layout and should be supported while importing, current issue is that

- is is not correctly imported as maven project
- it is not setting the BUNDLE_ROOT where the Manifest is located

@mickaelistria any idea why the source folders are not correctly imported?
I end up with the following `.classpath`:

```
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="src" path=""/>
	<classpathentry kind="output" path="target/classes"/>
</classpath>
```
and this project description:
```
<?xml version="1.0" encoding="UTF-8"?>
<projectDescription>
	<name>samemodule</name>
	<comment></comment>
	<projects>
	</projects>
	<buildSpec>
		<buildCommand>
			<name>org.eclipse.jdt.core.javabuilder</name>
			<arguments>
			</arguments>
		</buildCommand>
		<buildCommand>
			<name>org.eclipse.m2e.core.maven2Builder</name>
			<arguments>
			</arguments>
		</buildCommand>
		<buildCommand>
			<name>org.eclipse.pde.ManifestBuilder</name>
			<arguments>
			</arguments>
		</buildCommand>
		<buildCommand>
			<name>org.eclipse.pde.SchemaBuilder</name>
			<arguments>
			</arguments>
		</buildCommand>
	</buildSpec>
	<natures>
		<nature>org.eclipse.jdt.core.javanature</nature>
		<nature>org.eclipse.pde.PluginNature</nature>
		<nature>org.eclipse.m2e.core.maven2Nature</nature>
	</natures>
</projectDescription>
```

**Comment by mickaelistria** [comment]
This is not a specific import issue as later invocation of "Update Maven Project" doesn't seem to fix it either.
I suspect the unknown packaging can cause the configuration issue.

**Comment by laeubi** [comment]
I'm just wondering if we should be able to detect [Standard Directory Layout](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html) regardless of the packaging type?

**Comment by mickaelistria** [comment]
If we have enough context available, we shouldn't hardcode more the Standard Directory Layout than we already do; but it would be good fail back strategy to use in case we can't get such information from the context or if the MavenProject cannot be read to infer the layout from the context.
But this particular case is also a symptom of incomplete support for Maven extensions. If there is a good chance this gets working soon, it may not be worth implementing complexity here.

**Comment by laeubi** [comment]
I don't see how an extension would/should do anything... the standard layout is an layout maven understands with any configuration, any plugin so I think it is fair to assume m2e understand it out of the box.

**Comment by mickaelistria** [comment]
> I don't see how an extension would/should do anything...

The extension wouldn't per se; but supporting it would allow m2e to resolve the MavenProject, and thus to keep configuring the project normally instead of aborting like it seems to be doing.

> the standard layout is an layout maven understands with any configuration, any plugin so I think it is fair to assume m2e understand it out of the box.

It is already kind of supported for "sane" projects; and the issue here doesn't seem to be that this layout is not supported, but that m2e isn't able to properly configure projects that it cannot resolve to MavenProject because they require a Maven extension.
So adding support for the standard layout as failback would be some workaround, but not fixing the root cause.
IMO, instead of supporting the standard layout, it might make sense that if we detect an "unknown packaging type", we try to fail back to `jar` in the Maven build (but still show the error/warning for unknown packaging). Then the import would
work as expected; and more other things would.

**Comment by laeubi** [comment]
> The extension wouldn't per se; but supporting it would allow m2e to resolve the MavenProject, and thus to keep configuring the project normally instead of aborting like it seems to be doing.

Where do you see it anything aborting? THe project is imported, but the "src" folder is interpreted as the one and only source folder, something that is not reflected in the maven model where maven correctly understand this layout without any special action.

> It is already kind of supported for "sane" projects; and the issue here doesn't seem to be that this layout is not supported, but that m2e isn't able to properly configure projects that it cannot resolve to MavenProject because they require a Maven extension.

I don't see any error/exception related to extensions here, the packaging is correctly identified, you can try that out by removing the `<extension>true</extension>` where m2e will complain that the packaging type is unknown.

> it might make sense that if we detect an "unknown packaging type", we try to fail back to jar in the Maven build (but still show the error/warning for unknown packaging). Then the import would work as expected; and more other things would.

But that's exactly what is wrong here, the packaging type has nothing to do with the source folders (either automatically by convention or explicitly configured by the pom). One might argue that it is not fully clear if it is a **java project** but this is correctly identified and actually quite obvious by the `src/main/java` and `src/test/java` folders, so the `m2e.jdt` should actually always do something if:

- the maven model contains any source folders (that is by default  `src/main/java` and `src/test/java`)
- the project has a maven nature and thus one can have (java) source-folders.

The packaging type only tells what mojos are executed, and those mojos *might* read the declared source folders (e.g. maven-compiler-plugin), this is basically the same as with the `target` folder, no one would assume that different packaings would use different outputfolders if not explicitly configured....

**Comment by laeubi** [comment]
> Thinking further about it, the issue is most likely the absence of configured maven-compiler-plugin in the project definition; as another compiler plugin (tycho-compiler-plugin) is used.

But these folder are not from/for compiler plugin, so it should not be related to that plugin, these are standard maven model properties. So I think thats something that should be detected always, e.g. the resource folders are also not detected at all (and these releate to the packaging), otherwhise each packaging type must implement this over and over again what not seems very userful.

**Comment by mickaelistria** [comment]
>  these are standard maven model properties

Not really. Maven has a recommendation, but there is nothing that technically enforces it and it's easy for projects to use something else; that's why m2e decides -wisely- to check the project information (source/target directories, configuration of maven-compiler-plugin...) to configure the project instead of assuming -wrongly- everything obeys the Maven recommendation.
Again, trying the recommendation is a good fail-back strategy, but making it higher priority than checking the actual project as it's currently doing would cause some regression to every project that slightly disobeys that recommendation; and I've seen plenty of them...
